File: samples/java-micronaut/src/test/java/com/mongodb/client/reactor/sample/java/micronaut/SampleServiceTest.java
package com.mongodb.client.reactor.sample.java.micronaut;

import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoCollection;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import jakarta.inject.Inject;
import org.bson.types.ObjectId;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.testcontainers.containers.MongoDBContainer;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

@MicronautTest
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class SampleServiceTest implements TestPropertyProvider {

    private static final String DB_NAME = "test";

    private static final MongoDBContainer container = initContainer();

    @Inject
    private SampleService service;

    @Inject
    private MongoClient client;

    static MongoDBContainer initContainer() {
        MongoDBContainer container = new MongoDBContainer("mongo:4.4");
        container.start();
        return container;
    }

    @NotNull
    @Override
    public Map<String, String> getProperties() {
        return Collections.singletonMap("MONGO_PORT", container.getFirstMappedPort().toString());
    }

    @BeforeEach
    void setup() {
        MongoCollection<User> collection = client.getDatabase(DB_NAME).getCollection("users", User.class);
        Mono.from(collection.insertMany(Stream.of(
                new User().setFirstName("John").setLastName("Doe").setEmail("john.doe@mail.com"),
                new User().setFirstName("Jane").setLastName("Doe").setEmail("jane.doe@mail.com")
        ).collect(Collectors.toList()))).block();
    }

    @AfterEach
    void cleanUp() {
        Mono.from(client.getDatabase(DB_NAME).getCollection("users").drop()).block();
    }

    @Test
    void listDatabasesNamesShouldReturnTestDbName() {
        StepVerifier.create(service.listDatabasesNames())
                    .expectSubscription()
                    .recordWith(ArrayList::new)
                    .thenConsumeWhile(s -> true)
                    .consumeRecordedWith(l -> {
                        assertThat(l.size()).isGreaterThan(1);
                        assertThat(l).contains(DB_NAME);
                    })
                    .verifyComplete();
    }

    @Test
    void listCollectionsShouldReturnUsersCollection() {
        StepVerifier.create(service.listCollections())
                    .expectSubscription()
                    .expectNext("users")
                    .verifyComplete();
    }

    @Test
    void findAllShouldReturnMultipleUsers() {
        StepVerifier.create(service.findAll())
                    .expectSubscription()
                    .recordWith(ArrayList::new)
                    .expectNextCount(2)
                    .consumeRecordedWith(l -> assertThat(l.stream().map(User::getFirstName)).contains("John", "Jane"))
                    .verifyComplete();
    }

    @Test
    void createShouldAddNewDocument() {
        User user = new User().setFirstName("Mark").setLastName("Doe").setEmail("mark.doe@gmail.com").setId(new ObjectId());
        StepVerifier.create(service.create(user))
                    .expectSubscription()
                    .consumeNextWith(u -> assertThat(u).usingRecursiveComparison().isEqualTo(user))
                    .verifyComplete();
    }

}

File: samples/kotlin-micronaut/src/test/kotlin/com/mongodb/client/reactor/sample/kotlin/micronaut/SampleServiceTest.kt
package com.mongodb.client.reactor.sample.kotlin.micronaut

import com.mongodb.reactivestreams.client.MongoClient
import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import io.micronaut.test.support.TestPropertyProvider
import jakarta.inject.Inject
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.testcontainers.containers.MongoDBContainer
import reactor.kotlin.core.publisher.toMono
import reactor.kotlin.test.test

@MicronautTest
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
internal class SampleServiceTest : TestPropertyProvider {

    companion object {

        private const val DB_NAME = "test"
        private val container = MongoDBContainer("mongo:4.4").apply { start() }
    }

    @Inject lateinit var service: SampleService
    @Inject lateinit var client: MongoClient

    override fun getProperties() = mapOf("MONGO_PORT" to container.firstMappedPort.toString())

    @BeforeEach
    fun setup() {
        val collection = client.getDatabase(DB_NAME).getCollection("users", User::class.java)
        collection.insertMany(
            listOf(
                User("john.doe@mail.com", "John", "Doe"),
                User("jane.doe@mail.com", "Jane", "Doe")
            )
        ).toMono().block()
    }

    @AfterEach
    fun cleanUp() {
        client.getDatabase(DB_NAME).getCollection("users").drop().toMono().block()
    }

    @Test
    fun `list databases names should return test db name`() {
        service.listDatabasesNames().test()
            .expectSubscription()
            .recordWith { ArrayList() }
            .thenConsumeWhile { true }
            .consumeRecordedWith {
                assertThat(it.size).isGreaterThan(1)
                assertThat(it).contains(DB_NAME)
            }
            .verifyComplete()
    }

    @Test
    fun listCollectionsShouldReturnUsersCollection() {
        service.listCollections().test()
            .expectSubscription()
            .expectNext("users")
            .verifyComplete()
    }

    @Test
    fun findAllShouldReturnMultipleUsers() {
        service.findAll().test()
            .expectSubscription()
            .recordWith { ArrayList() }
            .expectNextCount(2)
            .consumeRecordedWith { s -> assertThat(s.stream().map { it.firstName }).contains("John", "Jane") }
            .verifyComplete()
    }

    @Test
    fun createShouldAddNewDocument() {
        val user = User("mark.doe@mail.com", "Mark", "Doe")
        service.create(user).test()
            .expectSubscription()
            .consumeNextWith { assertThat(it).usingRecursiveComparison().isEqualTo(user) }
            .verifyComplete()
    }
}
